# **This file works perfectly in google colab**
The Purpose of this doc is to calculate the next days Price given the Previous data.

In [1]:
import pandas as pd
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
import matplotlib

In [2]:
!pip install yfinance

  Created wheel for yfinance: filename=yfinance-0.1.54-py2.py3-none-any.whl size=22409 sha256=a182832da1dbf54e987dfb8930443e79f2fef239c2185eda32a409df5edd9670
  Stored in directory: /root/.cache/pip/wheels/f9/e3/5b/ec24dd2984b12d61e0abf26289746c2436a0e7844f26f2515c
Successfully built yfinance


In [3]:
import yfinance as yf

In [4]:
# Numerical libraries
from sklearn.model_selection import train_test_split

from sklearn.cluster import KMeans


Here to Predict Tommorows Price I have only included uptil January as taking too much of old data might reduce the effect of Post COVID pandemic

In [5]:
data = yf.download('TCS.NS', start = '2020-01-01',end = '2020-09-13')
data.head()

[*********************100%***********************]  1 of 1 completed


Open         High  ...    Adj Close   Volume
Date                                  ...                      
2020-01-01  2168.000000  2183.899902  ...  2135.918457  1354908
2020-01-02  2179.949951  2179.949951  ...  2126.113525  2380752
2020-01-03  2164.000000  2223.000000  ...  2168.484863  4655761
2020-01-06  2205.000000  2225.949951  ...  2168.287842  3023209
2020-01-07  2200.500000  2214.649902  ...  2173.609131  2429317

[5 rows x 6 columns]

In [6]:
y=[]
for i in range(1,len(data)):
  if data['Open'][i]-data['Close'][i-1]>0:
    y.append(1)#1 represents higher
  else:
    y.append(0)

In [7]:
x=data
x.drop(x.tail(1).index,inplace=True)

In [8]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.30, random_state=11)
model = LogisticRegression()
model.fit(x_train, y_train)
ypredict = model.predict(x_test)
pd.DataFrame(model.predict(x_train))
model.score(x_train,y_train)

0.6694214876033058

In [9]:
model.score(x_test, y_test)

0.6792452830188679

In [10]:
Algo = []
train = []
test = []
recall = []
roc = []

In [11]:
from sklearn.metrics import recall_score,precision_score,roc_auc_score
Algo.append('Logistic Regression')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))
recall.append(recall_score(ypredict,y_test,average = 'micro'))
roc.append(roc_auc_score(ypredict,y_test))

In [12]:
from sklearn.tree import DecisionTreeClassifier
model_decisiontree = DecisionTreeClassifier()
model_decisiontree.fit(x_train,y_train)
ypredict_decisiontree = model_decisiontree.predict(x_test)
print(model_decisiontree.tree_.max_depth)
model_decisiontree.score(x_train,y_train) 
model_decisiontree.score(x_test,y_test)

10


0.5471698113207547

In [13]:
from sklearn.tree import DecisionTreeClassifier
model_pruned = DecisionTreeClassifier(criterion='entropy',max_depth=3,random_state=5)
model_pruned.fit(x_train,y_train)
ypredict = model_pruned.predict(x_test)
Algo.append('Decision Tree using Entropy')
train.append(model_pruned.score(x_train,y_train))
test.append(model_pruned.score(x_test,y_test))
recall.append(recall_score(ypredict,y_test))
roc.append(roc_auc_score(ypredict,y_test))

In [14]:
model = DecisionTreeClassifier(criterion='gini',max_depth=3,random_state=5)
model.fit(x_train,y_train)
ypredict = model.predict(x_test)
Algo.append('Decision Tree using Gini')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))
recall.append(recall_score(ypredict,y_test))
roc.append(roc_auc_score(ypredict,y_test))

In [15]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators = 100)
model.fit(x_train,y_train)
ypredict = model.predict(x_test)
Algo.append('Random Forest')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))
recall.append(recall_score(ypredict,y_test))
roc.append(roc_auc_score(ypredict,y_test))

In [16]:
from sklearn.ensemble import BaggingClassifier
model = BaggingClassifier(n_estimators=100,max_samples=0.7,bootstrap=True,oob_score=True,random_state=11)
model.fit(x_train,y_train)
ypredict = model.predict(x_test)
Algo.append('Bagging model')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))
recall.append(recall_score(ypredict,y_test))
roc.append(roc_auc_score(ypredict,y_test))

In [17]:
from sklearn.ensemble import AdaBoostClassifier
model = AdaBoostClassifier(n_estimators=200,learning_rate=0.1,random_state=11)
model.fit(x_train,y_train)
ypredict = model.predict(x_test)
Algo.append('AdaBoost model')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))
recall.append(recall_score(ypredict,y_test))
roc.append(roc_auc_score(ypredict,y_test))

In [18]:
from sklearn.ensemble import GradientBoostingClassifier
model = GradientBoostingClassifier(n_estimators=200,random_state=11)
model.fit(x_train,y_train)
ypredict = model.predict(x_test)
Algo.append('Gradient Boosting model')
train.append(model.score(x_train,y_train))
test.append(model.score(x_test,y_test))
recall.append(recall_score(ypredict,y_test))
roc.append(roc_auc_score(ypredict,y_test))

In [19]:
results = pd.DataFrame()
results['Model'] = Algo
results['Training Accuracy'] = train
results['Testing Accuracy'] = test
results['Recall'] = recall
results['ROC AUC Score'] = roc
results = results.set_index('Model')
results

Training Accuracy  ...  ROC AUC Score
Model                                           ...               
Logistic Regression                   0.669421  ...       0.554167
Decision Tree using Entropy           0.710744  ...       0.480556
Decision Tree using Gini              0.719008  ...       0.480556
Random Forest                         1.000000  ...       0.507902
Bagging model                         0.983471  ...       0.518841
AdaBoost model                        0.859504  ...       0.586382
Gradient Boosting model               1.000000  ...       0.607771

[7 rows x 4 columns]

Here we Find with Adaboost we find the Maximum Accuracy of 67.9%

In [20]:
from pandas import DataFrame
y = DataFrame (y,columns=['Prediction'])

In [21]:
from keras.layers import LSTM,Dense,Dropout
from sklearn.model_selection import TimeSeriesSplit
from sklearn.metrics import mean_squared_error, r2_score
ts_split= TimeSeriesSplit(n_splits=5)
for train_index, test_index in ts_split.split(x):
  X_train, X_test = x[:len(train_index)], x[len(train_index): (len(train_index)+len(test_index))]
  y_train, y_test = y[:len(train_index)].values.ravel(), y[len(train_index): (len(train_index)+len(test_index))].values.ravel()

In [22]:
X_train =np.array(X_train)
X_test =np.array(X_test)

X_tr_t = X_train.reshape(X_train.shape[0], 1, X_train.shape[1])
X_tst_t = X_test.reshape(X_test.shape[0], 1, X_test.shape[1])

In [23]:
X_train = np.asarray(X_train).astype(np.float32)
y_train = np.asarray(y_train).astype(np.float32)
y_test = np.asarray(y_test).astype(np.float32)
X_tr_t = np.asarray(X_tr_t).astype(np.float32)
X_tst_t = np.asarray(X_tst_t).astype(np.float32)

In [24]:
from sklearn.metrics import mean_squared_error, r2_score
from keras.layers import LSTM,Dense,Dropout
from keras.models import Sequential
from keras.layers import Dense
import keras.backend as K
from keras.callbacks import EarlyStopping
from keras.optimizers import Adam
from keras.models import load_model
from keras.layers import LSTM
K.clear_session()
model_lstm = Sequential()
model_lstm.add(LSTM(16, input_shape=(1, X_train.shape[1]), activation='relu', return_sequences=False))
model_lstm.add(Dense(1))
model_lstm.compile(loss='mean_squared_error', optimizer='adam')
early_stop = EarlyStopping(monitor='loss', patience=5, verbose=1)
history_model_lstm = model_lstm.fit(X_tr_t, y_train, epochs=100, batch_size=8, verbose=1, shuffle=False)

Epoch 1/100
19/19 [==============================] - 0s 2ms/step - loss: 3496360.5000
Epoch 2/100
19/19 [==============================] - 0s 2ms/step - loss: 780107.6875
Epoch 3/100
19/19 [==============================] - 0s 2ms/step - loss: 53521.8984
Epoch 4/100
19/19 [==============================] - 0s 2ms/step - loss: 23085.9336
Epoch 5/100
19/19 [==============================] - 0s 2ms/step - loss: 9171.1797
Epoch 6/100
19/19 [==============================] - 0s 2ms/step - loss: 4637.8652
Epoch 7/100
19/19 [==============================] - 0s 2ms/step - loss: 4117.3511
Epoch 8/100
19/19 [==============================] - 0s 2ms/step - loss: 5036.3867
Epoch 9/100
19/19 [==============================] - 0s 2ms/step - loss: 6978.5659
Epoch 10/100
19/19 [==============================] - 0s 2ms/step - loss: 9838.3291
Epoch 11/100
19/19 [==============================] - 0s 1ms/step - loss: 13132.8984
Epoch 12/100
19/19 [==============================] - 0s 2ms/step - loss: 156

In [25]:
y_pred_test_lstm = model_lstm.predict(X_tst_t)
y_train_pred_lstm = model_lstm.predict(X_tr_t)
print("The R2 score on the Train set is:\t{:0.3f}".format(r2_score(y_train, y_train_pred_lstm)))
r2_train = r2_score(y_train, y_train_pred_lstm)
print("The R2 score on the Test set is:\t{:0.3f}".format(r2_score(y_test, y_pred_test_lstm)))
r2_test = r2_score(y_test, y_pred_test_lstm)


The R2 score on the Train set is:	-5.673
The R2 score on the Test set is:	-3.435
